### DISCUSSION TITLE
Create a vector_tile_layer with Maplibre Native Core

### DISCUSSION THREAD

**Discussion by T-moz** [fleaflet/flutter_map#1652] [Ideas]
I think that flutter_map provider the best devX to developper of any map sdk.

It's amazing, and we can build really cool stuff with it.

But right now the biggest limitation, is the support of vector tiles.
[flutter_map_vector_tile](https://github.com/greensopinion/flutter-vector-map-tiles) performance [issues](https://github.com/greensopinion/flutter-vector-map-tiles/issues?q=is%3Aissue++performance) seems to be caused by flutter rendering.

Instead we could use the [maplibre native C++ render](https://maplibre.org/maplibre-native/cpp/api/), and bind the controls to flutter_map.
Every custom  layers of Markers, polylines etc... would be rendered by flutter with flutter map, giving developers a lot of visual freedom, but the tiles will be rendered by maplibre.

**Comment by ibrierley** [comment]
Anyone is certainly welcome to create a proof of concept, it may give some ideas.

**Comment by JaffaKetchup** [comment]
It's also potentially worth noting that Impeller may drastically improve performance of the existing solutions, but I have checked myself.

**Comment by T-moz** [comment]
One of the benefits of using the maplibre renderer, will be sprits & text overlap and collision handling.
With impeller, we'll have to implement them from scratch.

Also, even with impeller enabled, the current implementation of [flutter_map_vector_tile](https://github.com/greensopinion/flutter-vector-map-tiles) is not as performant as maplibre.

**Comment by ibrierley** [comment]
I think it will be a case of both integration and maintenance look to be non-trivial. I think html has a different codebase as well. 
Hence a proof of concept would be good. I suspect this may be one of those which would need to be a plugin (but maybe there are some aspects that could be amended in fm to help the plugin work).

**Comment by T-moz** [comment]
Well I agree that it need to be a plugin.

The proof of concept can start with one or two platforms.

To create this render layer, I'm under the impression that multiple solutions are available, but I don't know witch one is best.

- Use the OG maplibre renderer.
- Use the new, cross-platform, rust based [maplibre renderer](https://github.com/maplibre/maplibre-rs).
- Copy all [glsl shaders](https://github.com/maplibre/maplibre-native/tree/main/shaders) from maplibre and use impeller with custom shaders.

**Comment by ibrierley** [comment]
Maybe this would be useful... https://github.com/maplibre/flutter-maplibre-gl as it uses flutter. I think there's some issues (erm looking at the issues), with things like Windows possibly not being supported.

**Reply by T-moz** [reply]
In fact I'm currently using [flutter-maplibre-gl](https://github.com/maplibre/flutter-maplibre-gl) and I will probably stick to it for now.
But, like google_map, and mapbox it has fondamental problems that flutter_map solves:

- The API is imperative and not modern.
- The platform view causes performance issues.
- It's not possible to display a flutter layer for markers or polylines with a position synchronise (without noticeable lag) to the tile layer.
- At leat with native sdk or react-native plugin, you can use view annotation, but it's not available in flutter.
